### PR TITLE
WIP: Move network-transport implemntation to the network-core

### PR DESCRIPTION
**Pull request by qnikst**
PR is done for an early review and has lots of work to do:

 * [ ] implement streaming replies
 * [ ] generalize messages, remove dependency on cardano
 * [ ] implement logging
 * [ ] implement subscription mechanism
 * [ ] verify that network can be injected into the loop in application
 * [ ] remove redundant constraints on the Node and it's associated types.

### CODE REVIEW

File: network-ntt/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+//!
+//! This module only provides and handle the different connections
+//! and act as message passing between the other modules (blockchain,
+//! transactions...);
+
+//extern crate cardano;
+//#[macro_use]
+//extern crate futures;
+//extern crate tokio;
+extern crate protocol_tokio as tokio_protocol;
```
Comment (mzabaluev) [inline network-ntt/src/lib.rs]: Just use `protocol` for brevity?

Comment (mzabaluev) [reply network-ntt/src/lib.rs]: Also, `extern crate` is unnecessary in Rust 2018.

File: network-ntt/src/reply_handle.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
+//! This module provides an abstraction to a simple connection
+//! reply mechanism. NT provides a very cheap lightweight connection,
```
Comment (mzabaluev) [inline network-ntt/src/reply_handle.rs]: I intended for `network-core` future and stream implementations to hide the details of serving the request. The `Server` implementation should be able to resolve requests with its embedded `Node` object.

File: network-ntt/src/server.rs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+//!                                     |      |
+//!                                     |      |-- connection1
+//!                                     |      |-- connection2
+//!  +--------------+     +----------+  |
+//!  | network-core | ----| NT state |--+---- server: sock2
+//!  +--------------+     +----------+  |      |
+//!                                     |      |-- connection3
+//!                                     |
+//!                                     +---- client: remote-sock1
+//! ```
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: Cool. Four-space indent should works too.

File: network-ntt/src/server.rs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+//! As NT state is shared all the services could provide a uniform
+//! access to the chain state. And one could build very flexible
+//! topology.
+//use crate::reply_handle::ReplyHandle;
+//use chain_core::property::{Deserialize, Serialize};
+use futures::prelude::*;
+use futures::{future, stream::{self,Stream}, sync::mpsc};
+use network_core::server;
+use network_core::server::block::BlockService;
+use network_core::server::transaction::TransactionService;
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: I tend to group imports by namespace with brackets, but I have no strong opinion about this.

File: network-ntt/src/server.rs [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+#[derive(Clone)]
+pub struct Server<N> {
+    node: N,
+}
+
+/// Run NTT server.
+pub fn run<N: 'static>(config: Config, node: N)
+where
+    N: std::marker::Sync + std::marker::Send + server::Node + Clone + Copy,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::TipFuture: std::marker::Send,
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: You can shorten the `network_core::server` paths to just `server`. Maybe it would be good to `use network_core::server as core_server` at the top for less ambiguity.

File: network-ntt/src/server.rs [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+        let listener =
+            stream::iter_ok(config.listen)
+                .for_each(move |sockaddr| {
+                    let server = Server {
+                          node: node.clone(),
+                    };
+                    run_listen_socket(sockaddr, server) 
+                });
+        let connections =
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: There should be a `Connection` object to encapsulate peer connection state, like in `network-grpc`

File: network-ntt/src/server.rs [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+// These types are loaded from cardano package,
+// because tokio-protocol is not generalized yet.
+// This should be fixed and then these types should go.
+use cardano::{
+    block::{HeaderHash},
+    tx::{TxAux},
+};
+
+/// Network transport server configuration.
+pub struct Config {
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: I did not bring a `Config` into the `network-grpc` library. Individual calls to `server::listen` and a future `server::connect` can be used to instantiate connections. I think we could live it up to the application how to manage collections of peers, unless we have a generally perfect idea for everyone.

File: network-ntt/src/server.rs [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::TipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::GetHeadersFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::StreamBlocksToTipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::TransactionService as network_core::server::transaction::TransactionService>::ProposeTransactionsFuture: std::marker::Send,
+{
+    let server = TcpListener::bind(&sockaddr)
+        .expect("Can't bind socket")
+        .incoming()
+        .map_err(move |_err| ())
+        .for_each(move |stream| {
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: I recommend separating out the listener stream creation to a helper function,
like done in `network-grpc`. The idea is for the application to be able to examine incoming connections as bare socket objects and decide if it wants to create protocol connections around them.

I will add another PR where the transport connection type is generalized in network-grpc, allowing TCP and Unix sockets. This is where such design also comes helpful.

File: network-ntt/src/server.rs [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+/// Internal structure of network transport node.
+#[derive(Clone)]
+pub struct Server<N> {
+    node: N,
+}
+
+/// Run NTT server.
+pub fn run<N: 'static>(config: Config, node: N)
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: This looks too monolithic to put it into a library.
The server application should be able to create or accept individual peer connections, and drive their futures in its preferred way. We will have connection task futures from both network-grpc and network-ntt at runtime, and we should be able to run them on the same thread pool for better resource utilization (if we ever manage to make them `Send`).

File: network-ntt/src/server.rs [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+                    tokio::spawn(run_connection(node, connection))
+                })
+        });
+    tokio::spawn(server)
+}
+
+/// Connect to another client.
+/// `run_connection` is spawned on the single heavyweight
+/// connection.
+pub fn run_connect_socket<N: 'static>(
```
Comment (mzabaluev) [inline network-ntt/src/server.rs]: I would put this away for another PR, when we design some sort of client API abstraction.

File: network-ntt/src/lib.rs (hunk 1 above)
Comment (qnikst) [reply network-ntt/src/lib.rs]: `protocol` clashes with `protocol` crate.

File: network-ntt/src/server.rs (hunk 10 above)
Comment (qnikst) [reply network-ntt/src/server.rs]: I'd keep it here due to the nature of the n-t, the client in network transport implementation will be some other thing.

File: network-ntt/src/lib.rs (hunk 1 above)
Comment (mzabaluev) [reply network-ntt/src/lib.rs]: That one is not imported, though? I don't think swapping words in the name brings more clarity to the code.

Comment (qnikst) [reply network-ntt/src/lib.rs]: Yeah that is fixed in the latest versions

File: network-ntt/src/server.rs (hunk 6 above)
Comment (qnikst) [reply network-ntt/src/server.rs]: Done

**Comment by mzabaluev** [comment]
I think we should close this as the work has been continued in #401. Move the TODO checklist there if it's still needed.
